interfaces/builtin/custom_device.go

```diff
@@ -185,22 +185,34 @@ func (iface *customDeviceInterface) validateUDevTaggingRule(rule map[string]inte
 			if err != nil {
 				break
 			}
-			deviceName := value.(string)
-			err = iface.validateKernelDeviceNameIsBasename(deviceName)
-			if err != nil {
-				break
-			}
-			// furthermore, the kernel name must match the basename of one
-			// of the given devices
+			kernelVal := value.(string)
+			// The kernel device name must match the full path of
+			// the given device, stripped of the leading /dev/, or
+			// it must be the basename of the device path.
 			foundMatch := false
 			for _, devicePath := range devices {
-				if deviceName == filepath.Base(devicePath) {
+				if "/dev/"+kernelVal == devicePath {
 					foundMatch = true
 					break
 				}
 			}
+			if foundMatch {
+				break
+			}
+			// Not a full path, so check if it matches the basename
+			// of a device path, and not more than one.
+			for _, devicePath := range devices {
+				if kernelVal != filepath.Base(devicePath) {
+					continue
+				}
+				if foundMatch {
+					err = fmt.Errorf(`%q matches more than one specified device`, kernelVal)
+					break
+				}
+				foundMatch = true
+			}
 			if !foundMatch {
-				err = fmt.Errorf(`%q does not match any specified device`, deviceName)
+				err = fmt.Errorf(`%q does not match any specified device`, kernelVal)
 			}
```

Comment:
This is a bit messy to have inside the for-loop, inside a switch statement. Can't we move this functionality to helpers (something like `validateUDevKernelDeviceName`)? This would make the for { switch {} } code easier to read.

Reply:
I completely agree. I've refactored it now, thanks for the suggestion!